config/config_helpers.go

```diff
@@ -0,0 +1,115 @@
+package config
+
+import (
+	"regexp"
+	"fmt"
+	"github.com/gruntwork-io/terragrunt/errors"
+	"github.com/gruntwork-io/terragrunt/options"
+	"path/filepath"
+	"github.com/gruntwork-io/terragrunt/util"
+)
+
+var INTERPOLATION_SYNTAX_REGEX = regexp.MustCompile("\\$\\{.*?\\}")
+var HELPER_FUNCTION_SYNTAX_REGEX = regexp.MustCompile("\\$\\{(.*?)\\(\\)\\}")
+var MAX_PARENT_FOLDERS_TO_CHECK = 100
+
+// Given a string value from a .terragrunt config file, parse the string, resolve any calls to helper functions using
+// the syntax ${...}, and return the final value.
+func ResolveTerragruntConfigString(terragruntConfigString string, parent *ParentConfig, terragruntOptions *options.TerragruntOptions) (resolved string, finalErr error) {
+	// The function we pass to ReplaceAllStringFunc cannot return an error, so we have to use named error
+	// parameters to capture such errors.
+	resolved = INTERPOLATION_SYNTAX_REGEX.ReplaceAllStringFunc(terragruntConfigString, func(str string) string {
+		out, err := resolveTerragruntInterpolation(str, parent, terragruntOptions)
+		if err != nil {
+			finalErr = err
+		}
+		return out
+	})
+
+	return
+}
+
+// Resolve a single call to an interpolation function of the format ${some_function()} in a .terragrunt file
+func resolveTerragruntInterpolation(str string, parent *ParentConfig, terragruntOptions *options.TerragruntOptions) (string, error) {
+	matches := HELPER_FUNCTION_SYNTAX_REGEX.FindStringSubmatch(str)
+	if len(matches) == 2 {
+		return executeTerragruntHelperFunction(matches[1], parent, terragruntOptions)
+	} else {
+		return "", errors.WithStackTrace(InvalidInterpolationSyntax(str))
+	}
+}
+
+// Execute a single Terragrunt helper function and return its value as a string
+func executeTerragruntHelperFunction(functionName string, parent *ParentConfig, terragruntOptions *options.TerragruntOptions) (string, error) {
+	switch functionName {
+	case "find_in_parent_folders": return findInParentFolders(terragruntOptions)
+	case "path_relative_to_parent": return pathRelativeToParent(parent, terragruntOptions)
+	default: return "", errors.WithStackTrace(UnknownHelperFunction(functionName))
+	}
+}
+
+// Find a parent .terragrunt file in the parent folders above the current .terragrunt file and return its path
+func findInParentFolders(terragruntOptions *options.TerragruntOptions) (string, error) {
+	previousDir := filepath.Dir(terragruntOptions.TerragruntConfigPath)
+
+	// To avoid getting into an accidental infinite loop (e.g. do to cyclical symlinks), set a max on the number of
+	// parent folders we'll check
+	for i := 0; i < MAX_PARENT_FOLDERS_TO_CHECK; i++ {
+		currentDir := filepath.Dir(previousDir)
+		if currentDir == previousDir {
+			return "", errors.WithStackTrace(ParentTerragruntConfigNotFound(terragruntOptions.TerragruntConfigPath))
+		}
```

Comment:
This cycle check will miss any symlink that isn't in the immediate parent folder. Should you instead maintain a slice (or map) of paths and check against that?